### DISCUSSION TITLE
hands-on-code/from-self-attention-to-multi-head-self-attention

### DISCUSSION THREAD

**Discussion by giscus** [bbruceyuan/bbruceyuan.github.io#21] [Announcements]
# hands-on-code/from-self-attention-to-multi-head-self-attention

在 AI 相关的面试中，经常会有面试官让写 self-attention，但是 transformer 这篇文章其实包含很多的细节，因此可能面试官对于 self-attention 实现到什么程度是有不同的预期。因此这里想通过写不同版本的 self-attention 实现来达到不同面试官的预期，四个不同的版本，对应不同的细节程度。

https://yuanchaofa.com/hands-on-code/from-self-attention-to-multi-head-self-attention

**Comment by Abigail61** [comment]
output_proj也不是multihead的产物，在self attention中也可以理解。
自然理解是，在Self-Attention中，每个词向量E经过Attention层后会得到一个偏移量△E。这个偏移量△E会通过残差连接（residual connection）与原始词向量E相加，用来更新词向量的表示。所以为了确保△E的维度与E一致，W_v矩阵的输出维度需要与embedding_size相同，而不是hidden_size。

但是，如果直接将W_v矩阵的输出维度设置为embedding_size，由于embedding_size >> hidden_size， 所以W_v矩阵会变得非常大，导致计算复杂度增加。为了减少参数量和计算量，可以将W_v矩阵分解为两个部分：W_V1和W_V2。维度为 embedding_size * hidden_size, hiddensize_ embedding_size, 其中，W_V1就是通常所说的W_v矩阵，而W_V2则作为输出投影矩阵（o_proj矩阵）

**Reply by bbruceyuan** [reply]
嗯嗯，可以这么解释，这个解释挺好的；但是一般来说，self attention的公式不包括 o_proj

**Reply by NEMOlv** [reply]
博主这里没有考虑非整除的情况。实际上不管是huggingface attn实现还是pytorch MHA的实现，qkv的input_size=hidden_dim，output_size=head_dim * nums_head，而o_proj的input_size和output_size与qkv相反。o_proj这里还起到一个在非整除情况下恢复原始hidden_dim的作用。

**Comment by qigemingzifeibantianjin** [comment]
感谢分享！不过attention_mask的第二个向量不要设置为【0,0】，会让batch的第二个seq无效。

**Reply by bbruceyuan** [reply]
嗯嗯理解你这个意思，这里的 attention mask 就是随便写的，想说明哪怕一个 sample 全都是 pad token 也没有问题。（针对于：sample 凑不齐一个 batch，但是又不想舍弃任何sample，会存在这样的情况）
